### DISCUSSION TITLE
Query for java.time.LocalDateTime type is possible?

### DISCUSSION THREAD

**Discussion by naeul-lee** [Netflix/dgs-framework#844] [Q&A]
Hi community!
I'm working on some project where I query data from MySQL (using QueryDSL),
then I want serving this data using dgs framework.

I fell into the problem that I want to keep my entity class have "java.time.LocalDateTime" type,
and serving this data to my clients.

but failing like below

"Can't serialize value (/summary[0]/eventAt) : Expected something we can convert to 'java.time.OffsetDateTime' but was 'LocalDateTime'

![image](https://user-images.githubusercontent.com/24990701/151115208-93349191-6473-4867-a5c2-4a93ec1c13ac.png)

----
This is my dto class

class Summary @QueryProjection constructor(
    val ruleSetId: Long?,
    val ruleSetName: String?,
    val count: Long,
    val eventAt: LocalDateTime
)

---- 
This is my schema

type Summary {
  ruleSetId: ID
  ruleSetName: String
  count: Int
  eventAt: DateTime
}

----

scalar DateTime

I added DateTimeScalar class (referenced https://netflix.github.io/dgs/scalars/)

`
@DgsScalar(name = "DateTime")
class DateTimeScalar : Coercing<LocalDateTime, String> {
    @Throws(CoercingSerializeException::class)
    override fun serialize(dataFetcherResult: Any): String {
        return if (dataFetcherResult is LocalDateTime) {
            dataFetcherResult.format(DateTimeFormatter.ISO_DATE_TIME)
        } else {
            throw CoercingSerializeException("Not a valid DateTime")
        }
    }

    @Throws(CoercingParseValueException::class)
    override fun parseValue(input: Any): LocalDateTime {
        return LocalDateTime.parse(input.toString(), DateTimeFormatter.ISO_DATE_TIME)
    }

    @Throws(CoercingParseLiteralException::class)
    override fun parseLiteral(input: Any): LocalDateTime {
        if (input is StringValue) {
            return LocalDateTime.parse(input.value, DateTimeFormatter.ISO_DATE_TIME)
        }
        throw CoercingParseLiteralException("Value is not a valid ISO date time")
    }
}

`

**Comment by srinivasankavitha** [comment]
Are you able to verify that the custom scalar serialize/deserialization is
being invoked properly? You could try setting a breakpoint there to check.
If it is not, the scalar is not wired up properly. If you see that it is
being invoked, we'd have to look at the data fetcher implementation to
investigate further.

**Reply by naeul-lee** [reply]
Thank you for quick response !

@srinivasankavitha 
It seems like the scalar serialize / deserialize is not invoked.
I set breakpoints all around the "DateTimeScalar" class, but nothing is captured.


as @berngp suggested, 
I understand the "DateTimeScalar" is another way to deal with custom scala type other than "extended scalars".
So I removed all the code and dependecies about `com.graphql-java:graphql-java-extended-scalars`.


But I also tried including extended scalars and used like below, 
But this is not working either.

@DgsComponent
class DateTimeScalarRegistration {

    @DgsRuntimeWiring
    fun addScalar(builder: RuntimeWiring.Builder): RuntimeWiring.Builder {
        return builder.scalar(ExtendedScalars.DateTime)
    }
}

**Comment by berngp** [comment]
As @srinivasankavitha suggests, looks like the extended scalars were not registered. Have you tried just importing the `com.graphql-java:graphql-java-extended-scalars` dependency? Ref [register-scalar-extensions-via-dgsruntimewiring](https://netflix.github.io/dgs/scalars/#register-scalar-extensions-via-dgsruntimewiring).

**Comment by srinivasankavitha** [comment]
Could you please share your code or some example to reproduce this issue?
That will help us understand how you are setting it up.

**Reply by Cuica20** [reply]
my example:

`package com.dsrp.backend.dgs;

import com.netflix.graphql.dgs.DgsScalar;
import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.CoercingParseLiteralException;
import graphql.schema.CoercingParseValueException;
import graphql.schema.CoercingSerializeException;

import java.sql.Timestamp;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

@DgsScalar(name="Date")
public class DateScalar implements Coercing<LocalDate, String> {
    @Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
        if (dataFetcherResult instanceof Timestamp) return dataFetcherResult.toString();
        else throw new CoercingSerializeException("Expected a Timestamp object.");
    }

    @Override
    public LocalDate parseValue(Object input) throws CoercingParseValueException {
        try {
            if (input instanceof String) return LocalDate.parse((String) input, DateTimeFormatter.ofPattern("dd-MM-yyyy"));
            else throw new CoercingParseValueException("Expected a String");
        } catch (DateTimeParseException e) {
            throw new CoercingParseValueException(String.format("Not a valid date: %s", input), e);
        }
    }

    @Override
    public LocalDate parseLiteral(Object input) throws CoercingParseLiteralException {
        if (input instanceof StringValue) {
            try {
                return LocalDate.parse(((StringValue) input).getValue());
            } catch (DateTimeParseException e) {
                throw new CoercingParseLiteralException();
            }
        } else throw new CoercingParseLiteralException("Expected a StringValue");
    }
}

`

in my schema

type Module {
    moduleId: ID!
    orderByCourse: Int!
    name: String!
    quantity: Int!
    course: Course
    recordedClass: [RecordedClass]
    resource:[Resource]
    quiz: [Quiz]
    createdAt: Date!
    updatedAt: Date!
}
